src/blueapi/service/model.py

```diff
@@ -40,10 +42,35 @@ def from_device(cls, device: Device) -> "DeviceModel":
         name = device.name if isinstance(device, HasName) else _UNKNOWN_NAME
         return cls(name=name, protocols=list(_protocol_info(device)))
 
+    @classmethod
+    def from_device_tree(cls, root: Device, max_depth: int) -> list["DeviceModel"]:
```

Comment:
Should: There is a function in `device_lookup.py` that also traverses the device tree, could extract the walk logic into a generator and use it in both of these.